smartsim/launchable/job.py

```diff
@@ -24,13 +24,18 @@
 # OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
+from __future__ import annotations
+
+import typing as t
 from copy import deepcopy
 
 from smartsim._core.commands.launchCommands import LaunchCommands
-from smartsim.entity.entity import SmartSimEntity
 from smartsim.launchable.basejob import BaseJob
 from smartsim.settings import LaunchSettings
 
+if t.TYPE_CHECKING:
```

Comment:
Why is SmartSimEntity conditional now?

Reply:
It's there to prevent a circular import from occurring!

<details>
<summary> Info dump if you're curious about details</summary>

One of the problems with python is that it does not have a "true" compile step, and therefore doesn't have header files or really anyway to separate declaration from implementation. Usually this is a good thing IMO, but it becomes kind of a pain when type hinting python specifically.

For example consider the following modules:
```py
# my_module_a.py
from my_module_b import B

class A:
    def do_something_with_a_b(self, b: B) -> None: ...
```
```py
# my_module_b.py
from my_module_a import A

class B:
    def do_something_with_an_a(self, a: A) -> None: ...
```
and
```py
# main.py
from my_module_a import A
from my_module_b import B

def main():
    a = A()
    b = B()
    a.do_something_with_a_b(b)
    b.do_something_with_an_a(a)

if __name__ == "__main__":
    raise SystemExit(main())
```
We can see that this _should_ work, but when we actually try to run `python main.py` we get the following error:
```py
Traceback (most recent call last):
  File "/private/tmp/d/main.py", line 2, in <module>
    from my_module_a import A
  File "/private/tmp/d/my_module_a.py", line 2, in <module>
    from my_module_b import B
  File "/private/tmp/d/my_module_b.py", line 2, in <module>
    from my_module_a import A
ImportError: cannot import name 'A' from partially initialized module 'my_module_a' (most likely due to a circular import) (/private/tmp/d/my_module_a.py)
```
we can see that `my_module_a` requires `my_module_b` which requires `my_module_a` which leads to a circular dependency at run time.

BUT! We don't _actually_ need the dependency at runtime, we just need it for type checking, so by moving to imports into a `t.TYPE_CHECKING` block and changing the type hints to forward references(i.e. strings (which is what `from __future__ import annotations` does behind the scenes (for now ([see here if you're curious](https://docs.python.org/3/library/__future__.html#id2))))), we can use their declaration for static analysis, but ignore their implementation at run time.
```py
# my_module_a.py (revised)
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from my_module_b import B

class A:
    def do_something_with_a_b(self, b: "B") -> None: ...
    #             Note the forward ref ^^^
```
```py
# my_module_b.py (revised)
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from my_module_a import A

class B:
    def do_something_with_an_a(self, a: "A") -> None: ...
    #              Note the forward ref ^^^
```

and now running `python main.py` works as expected!
</details>

Reply:
thanks!